add support for dbcp 2.x plugin

Jargon96 opened the pull request:
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- This plugin can collect the information of dbcp

#### What processes can this plugin monitor?
- get connection from pool
- close connection from pool
- create pool
- close pool
![image](https://user-images.githubusercontent.com/40067375/94046953-f232e480-fe03-11ea-81e4-06aede61f687.png)



#### Other Information
There are something wrong with mysql plugin, I have issued this problem in [5549](https://github.com/apache/skywalking/issues/5549)
![image](https://user-images.githubusercontent.com/40067375/94047509-c82df200-fe04-11ea-99e8-25fbbee8c36b.png)

ascrutae commented:
CI has broken.  You'd better to fix it.

Jargon96 commented:
> CI has breaking. You'd better to fix it.

Thank you sir.I have updated the code.

Jargon96 commented:
updated plugin example，This is the statement I executed：
* create table test_DBCP
* drop table test_DBCP
After the statement is executed, I closed the connection and returned it to the connection pool
![image](https://user-images.githubusercontent.com/40067375/94346608-5b169880-0060-11eb-8dbe-b1d141bb0847.png)

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/src/main/java/org/apache/skywalking/apm/testcase/dbcp/SQLExecutor.java:
```diff
@@ -0,0 +1,85 @@
+    public void createTable(String sql) throws SQLException {
+        Statement statement = connection.createStatement();
+        statement.execute(sql);
+        statement.close();
```
This is not the right way to close `Statement`.

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/src/main/java/org/apache/skywalking/apm/testcase/dbcp/SQLExecutor.java:
```diff
@@ -0,0 +1,85 @@
+            properties.setProperty("url", MysqlConfig.getUrl());
+            properties.setProperty("username", MysqlConfig.getUserName());
+            properties.setProperty("password", MysqlConfig.getPassword());
+            ds = BasicDataSourceFactory.createDataSource(properties);
```
Why init the static field in the constructor?

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/src/main/java/org/apache/skywalking/apm/testcase/dbcp/SQLExecutor.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public void close() throws Exception {
+        closeConnection();
+        closePool();
```
Don't use the closing pool. The pool should have the same life cycle as the application.

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/PoolingCloseConnectInterceptor.java:
```diff
@@ -0,0 +1,53 @@
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        AbstractSpan span = ContextManager.createLocalSpan("DBCP/Connection/" + method.getName());
+        span.setComponent(ComponentsDefine.DBCP);
+        SpanLayer.asDB(span);
```
Connection pool is not a DB layer.

Jargon96 replied:
Dear Mr Wu @wu-sheng  ,sorry to bother you, but could you give me some help?

kezhenxu94 replied:
> Dear Mr Wu @wu-sheng ,sorry to bother you, but could you give me some help?

He is on vocation recently.

From the functionality of connection pool, I think it's ok to consider it as cache layer.

Jargon96 replied:
> > Dear Mr Wu @wu-sheng ,sorry to bother you, but could you give me some help?
> 
> He is on vocation recently.
> 
> From the functionality of connection pool, I think it's ok to consider it as cache layer.
I got it！thanks for your reply！

wu-sheng replied:
We don't.require layer for all plugins. For the connection pool, layer is not required.

Jargon96 replied:
> We don't.require layer for all plugins. For the connection pool, layer is not required.

Thank you sir, and wish you a happy vacation!

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/PoolingCloseConnectInterceptor.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        if (ContextManager.isActive()) {
```
Why need this?

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/PoolingCloseConnectInterceptor.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Throwable t) {
+        if (ContextManager.isActive()) {
```
Why need this?

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/PoolingGetConnectInterceptor.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.lang.reflect.Method;
+
+public class PoolingGetConnectInterceptor implements InstanceMethodsAroundInterceptor {
```
Take the reference from the review on `PoolingCloseConnectInterceptor`.

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/PoolingCloseConnectInterceptor.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.lang.reflect.Method;
+
+public class PoolingCloseConnectInterceptor implements InstanceMethodsAroundInterceptor {
```
Comments are required.

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/define/BasicDataSourceInstrumentation.java:
```diff
@@ -0,0 +1,71 @@
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+public class BasicDataSourceInstrumentation extends ClassEnhancePluginDefine {
```
Please extends `ClassInstanceMethodsEnhancePluginDefine`.

wu-sheng replied:
Comments required.

wu-sheng commented on apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/define/DelegatingConnectionInstrumentation.java:
```diff
@@ -0,0 +1,71 @@
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+public class DelegatingConnectionInstrumentation extends ClassEnhancePluginDefine {
```
Please extends `ClassInstanceMethodsEnhancePluginDefine`.

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/src/main/java/org/apache/skywalking/apm/testcase/dbcp/controller/CaseController.java:
```diff
@@ -0,0 +1,58 @@
+    @ResponseBody
+    public String testcase() throws Exception {
+        try (SQLExecutor sqlExecute = new SQLExecutor()) {
+            sqlExecute.createTable(CREATE_TABLE_SQL);
+            sqlExecute.dropTable(DROP_TABLE_SQL);
```
Create and drop are not typical SQLs in the execution stage, insert/update/delete are.

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/src/main/java/org/apache/skywalking/apm/testcase/dbcp/controller/CaseController.java:
```diff
@@ -0,0 +1,58 @@
+    @RequestMapping("/dbcp-2.x-scenario")
+    @ResponseBody
+    public String testcase() throws Exception {
+        try (SQLExecutor sqlExecute = new SQLExecutor()) {
```
You are using `SQLExecutor` like a connection, but in the codes, it actually is controlling the whole datasource. This is not right. The real services don't use it in this way.

wu-sheng reviewed (changes requested):
`Supported-list.md` doc is required to update. The test codes are still my concern.

Also, @dmsolr, as the mentor, please review the codes first

Jargon96 commented:
I updated plugin and modified some inappropriate places, please review.

This is the statement I executed in test case：
* CREATE TABLE test_DBCP
* INSERT INTO test_DBCP(id, value) VALUES(1,1)
* SELECT id, value FROM test_DBCP WHERE id=1
* DELETE FROM test_DBCP WHERE id=1
* DROP table test_DBCP
After the statement is executed, I closed the connection and returned it to the connection pool.
![test_case](https://user-images.githubusercontent.com/40067375/97024987-1102d300-158a-11eb-9a4c-48a11f8991b8.png)

wu-sheng commented:
@Jargon96  Is @ascrutae your mentor? Or someone else?

Jargon96 commented:
> @Jargon96 Is @ascrutae your mentor? Or someone else?

@wu-sheng No,sir. @dmsolr is my mentor.

wu-sheng commented:
Please ping @dmsolr, he has enough experience in plugin development. I prefer he could review first.

Jargon96 commented:
> Please ping @dmsolr, he has enough experience in plugin development. I prefer he could review first.

@wu-sheng Got it, thank you sir.
@dmsolr Sir，could you help me review code?

dmsolr commented on test/plugin/scenarios/dbcp-2.x-scenario/config/expectedData.yaml:
```diff
@@ -0,0 +1,255 @@
+segmentItems:
+- serviceName: dbcp-2.x-scenario
+  segmentSize: nq 0
+  segments:
+  - segmentId: not null
+    spans:
+    - operationName: /dbcp-2.x-scenario/case/healthCheck
+      operationId: eq 0
+      parentSpanId: -1
+      spanId: 0
+      spanLayer: Http
+      componentId: 1
+      isError: false
+      spanType: Entry
+      peer: ''
+      skipAnalysis: false
+      tags:
+        - {key: url, value: 'http://localhost:8080/dbcp-2.x-scenario/case/healthCheck'}
+        - {key: http.method, value: HEAD}
```
we don't need to validate this segment.

Jargon96 replied:
> we don't need to validate this segment.

@dmsolr Thank you sir, I have deleted this segment.

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/support-version.list:
```diff
@@ -0,0 +1,20 @@
+2.7.0
+2.6.0
+2.5.0
+2.4.0
```
Do DBCP have 2.0-2.4?

Jargon96 replied:
> Do DBCP have 2.0-2.4?

@wu-sheng Thanks for your review~ Yes, sir.  It also has these versions
2.0 - 2.4: 2.3.0/2.2.0/2.1.1/2.1/2.0.1/2.0 
I've tested these versions locally. Do I need to add it to the list？

wu-sheng replied:
We should add the latest of each version, 2.3.0, 2.2.0, 2.1.1, 2.0.1

Jargon96 replied:
> We should add the latest of each version, 2.3.0, 2.2.0, 2.1.1, 2.0.1

Ok sir, I have updated the list.

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/config/expectedData.yaml:
```diff
@@ -0,0 +1,240 @@
+        - key: db.statement
+          value: "CREATE TABLE test_DBCP(\nid VARCHAR(1) PRIMARY KEY, \nvalue VARCHAR(1)\
+          \ NOT NULL)"
+    - operationName: DBCP/Connection/close
```
@Jargon96 @dmsolr @ascrutae Does DBCP pool have a name? I hope this could be `DBCP/{pool-name}/...`. Then some of the operations like `get` and `close` should be the logic span later to show the connection pool performance.

Jargon96 replied:
> @Jargon96 @dmsolr @ascrutae Does DBCP pool have a name? I hope this could be `DBCP/{pool-name}/...`. Then some of the operations like `get` and `close` should be the logic span later to show the connection pool performance.

@wu-sheng Could I use this ?  `DBCP/Apache-Commons-DBCP/...`

wu-sheng replied:
See my comment below.

dmsolr replied:
@wu-sheng  said whether the pool has something like a name or others to identify it.
As far as I see, it does not have.

ascrutae replied:
I checked the API and found that DBCP doesn't have attribute `name` to set pool name.

Jargon96 replied:
@dmsolr @ascrutae Thank you for checking. I didn’t find it either.

wu-sheng replied:
OK, then let's keep the name as what it is today.

wu-sheng commented on test/plugin/scenarios/dbcp-2.x-scenario/src/main/java/org/apache/skywalking/apm/testcase/dbcp/service/CaseService.java:
```diff
@@ -0,0 +1,71 @@
+    private static final String DROP_TABLE_SQL = "DROP table test_DBCP";
+
+    static {
+        Properties properties = new Properties();
+        properties.setProperty("driverClassName", "com.mysql.jdbc.Driver");
+        properties.setProperty("url", MysqlConfig.getUrl());
+        properties.setProperty("username", MysqlConfig.getUserName());
+        properties.setProperty("password", MysqlConfig.getPassword());
+        try {
+            ds = BasicDataSourceFactory.createDataSource(properties);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```
I mean, does the connection pool has an attribute called `name` or something?

wu-sheng replied:
If not, then there is nothing to be changed.

wu-sheng reviewed (approved):
LGTM. @ascrutae @dmsolr Please confirm.

ascrutae reviewed (approved):
LGTM

dmsolr reviewed (dismissed):
LGTM

wu-sheng commented:
@Jargon96 I think you have to modify the shardingsphere's expected data file. Because you added the DBCP plugin, and they use it in their plugin too. As this is a new default plugin, their expected files changed.

![image](https://user-images.githubusercontent.com/5441976/98065354-6ba40500-1e8f-11eb-8af8-749400594fcd.png)

Their codes are 
![image](https://user-images.githubusercontent.com/5441976/98065372-73fc4000-1e8f-11eb-94a4-8e697f4a8b0e.png)
Ping @yu199195 if you need some guidance.

wu-sheng reviewed (changes requested):
Just wait for the ShardingSphere test case fix.

Jargon96 commented:
> @Jargon96 I think you have to modify the shardingsphere's expected data file. Because you added the DBCP plugin, and they use it in their plugin too. As this is a new default plugin, their expected files changed.
> 
> ![image](https://user-images.githubusercontent.com/5441976/98065354-6ba40500-1e8f-11eb-8af8-749400594fcd.png)
> 
> Their codes are
> ![image](https://user-images.githubusercontent.com/5441976/98065372-73fc4000-1e8f-11eb-94a4-8e697f4a8b0e.png)
> Ping @yu199195 if you need some guidance.

@wu-sheng Ok sir, I'll try to modify it  :)

wu-sheng commented:
@Jargon96 The ShardingSphere plugin has fixed their plugin test w/o the DBCP dependency. Please resolve the conflicts and make your test passed. Other cases should be not the block for you.

Jargon96 commented:
@wu-sheng Thank you sir, I have updated the code.

kezhenxu94 reviewed (commented):
Seems you forgot to update this https://github.com/apache/skywalking/blob/master/docs/en/setup/service-agent/java-agent/Supported-list.md

Jargon96 commented:
> Seems you forgot to update this https://github.com/apache/skywalking/blob/master/docs/en/setup/service-agent/java-agent/Supported-list.md

Dear sir, Is cache classification appropriate？or create a new classification?

wu-sheng commented:
> Dear sir, Is cache classification appropriate？or create a new classification named DBCP?

Add `Pool` catalog.

Jargon96 commented:
> Add `Pool` catalog.

Got it.

wu-sheng reviewed (dismissed):
LGTM. @kezhenxu94 Any other suggestion?

kezhenxu94 reviewed (dismissed):
LGTM

wu-sheng commented:
One last thing, `CHANGES.md` is not updated. You are adding a new plugin.

Jargon96 commented:
> One last thing, `CHANGES.md` is not updated. You are adding a new plugin.

Ok sir, I have updated CHANGES.md.

wu-sheng reviewed (approved):
@Jargon96  
Congrats on finished **SUMMER 2020 project** officially. This is not required by the project committee, but as the SkyWalking community, we expect you could go through this process. YOU MADE IT!
This is how open source really works in the real world. The community is being welcome to new contributors and very strict about codes at the same time. I know this was very frustrating at the beginning, and you had to fix many issues and follow the comments of reviewers.

I hope this moment(`merging`) could make you happy eventually as be the 281st contributor to this main repo. Hope this journey would help you more in the future of your software engineering work. Eventually, for now, open-source is leading the whole software industry.

kezhenxu94 reviewed (approved):
Thanks for your patience

Jargon96 commented:
I was very excited about the moment of code merging! !  I'm inspired and proud to be  the 281st contributor to the main repo!

Thanks to the SkyWalking community, especially my mentor @dmsolr , Mr Wu @wu-sheng , Mr.Ke @kezhenxu94 and Mr.Zhang @ascrutae .This journey is meaningful to me. The SkyWalking community is awesome.  Participating in open source is a really interesting and cool thing!  I want to keep going deeper : )

Again, thank you for your patient guidance!
